File: src/CoinChange2.java
//                              Top-down                            Bottom-Up
// Time Complexity:     O(amount/minimum_denomination)               O(m*n)
// Space Complexity:    O(2^(amount/minimum_denomination))            O(1)
// Yes, this code ran successfully
// No, I didn't face any problem in this problem statement

package src;

import java.util.Arrays;

public class CoinChange2
{
    public static void main(String[] args) {
        int[] coins = {186,419,83,408};
        int amount = 6249;
        Change2 cs = new Change2();
        System.out.println(cs.change(amount,coins));
    }
}

class Change2
{
    int[][] dp;
    public int change(int amount, int[] coins) {
        int n = coins.length;
        dp = new int[n][amount+1];
        for(int i=0; i<n; i++){
            for(int j=0; j<amount+1; j++)
            {
                dp[i][j] = -1;
            }
        }
        return change(coins, 0, amount);

// ***************************** Bottom-Up Approach *****************************
//        long[] dp = new long[amount+1];
//        Arrays.fill(dp, 0);
//        dp[0] = 1;
//        for(int i=0; i<coins.length; i++)
//        {
//            int coin=coins[i];
//            for(int j=0; j<=amount; j++)
//            {
//                dp[j] = (j<coin) ? dp[j] : dp[j-coin]+dp[j] ;
//            }
//        }
//        return dp[amount]==0 ? 0 : (int)dp[amount];

    }
    private int change(int[] coins, int index, int amount)
    {
        if(amount == 0)                                                             // if amount is made out of this way
            return 1;
        if(amount < 0 || index == coins.length)                                     // if amount can't be made out of this way
            return 0;
        if(dp[index][amount] >= 0)
            return dp[index][amount];
        int noChoose = change(coins, index+1, amount);                        // not considering current coin
        int choose = change(coins, index, amount-coins[index]);             // considering current coin

        return  dp[index][amount] = noChoose+choose;                                // Getting number of ways to get amount
    }
}

File: Coin_Change-2.cpp
//Time complexity: O(N*M) , N number of houses
//Space complexity: O(1)

#include <iostream>
#include <vector>

using namespace std;
class Solution {
public:
     int change(int amount, vector<int>& coins) {
        //Using 1D array
          vector<int> dp(amount+1,0);
        // int* dp=new int[amount+1];
        // for(int j=0;j<amount+1;j++){ dp[j]=0;}
         dp[0]=1;   //using zero coin,  zero amount can be abtained in one way
         for(int i=1;i<coins.size()+1;i++){
             for(int j=0;j<amount+1;j++){
                 if(j>=coins[i-1])
                     dp[j]=dp[j]+dp[j-coins[i-1]];
             }
         }
         return dp[amount];
         
         
         
        /*
        using 2D array
        vector<vector<int>> dp(coins.size() + 1, vector<int>(amount + 1,0));
        for (int i = 0;i < coins.size() + 1;i++)
            dp[i][0] = 1;
        for (int i = 1;i < coins.size() + 1;i++) {
            for (int j = 1;j < amount + 1;j++) {
                if (j < coins[i - 1]) {
                    dp[i][j] = dp[i - 1][j];
                }
                else {
                    dp[i][j] = dp[i - 1][j] + dp[i][j - coins[i - 1]];

                }
            }
        }
        return dp[coins.size()][amount];*/
    }
};

int main(int argc, const char * argv[]) {
    // insert code here...
    vector<int> nums{1,2,5};
    int amount=7;
    Solution s;
    cout << s.change(amount, nums) << "\n";
    return 0;
}
